### ISSUE TITLE
Missing duplicate flag field in message return from mqtt-in node

### ISSUE THREAD

**Issue by ledat1011** [node-red/node-red#3495]
### Current Behavior

When using `mqtt-in` node the message return from this node missing duplicate flag field in message return.
According mqtt library nodejs, it provide duplicate flag (dup) field in message when subscribe topic from broker.
I think you also using this library to handle mqtt client, so could you help me add this field to the message return from `mqtt-in` node so that I can handle some logic on it.

### Expected Behavior

The message return from `mqtt-in` node must be included duplicate flag field.

### Example flow

```
paste your flow here
```
[
    {
        "id": "3af82246.3634ae",
        "type": "tab",
        "label": "Flow 1",
        "disabled": false,
        "info": ""
    },
    {
        "id": "9bc7090c819f05c1",
        "type": "debug",
        "z": "3af82246.3634ae",
        "name": "",
        "active": true,
        "tosidebar": true,
        "console": false,
        "tostatus": false,
        "complete": "true",
        "targetType": "full",
        "statusVal": "",
        "statusType": "auto",
        "x": 630,
        "y": 200,
        "wires": []
    },
    {
        "id": "c703697adf79fe86",
        "type": "mqtt in",
        "z": "3af82246.3634ae",
        "name": "",
        "topic": "test",
        "qos": "2",
        "datatype": "auto",
        "broker": "4c2c9454f646f701",
        "nl": false,
        "rap": true,
        "rh": 0,
        "inputs": 0,
        "x": 410,
        "y": 200,
        "wires": [
            [
                "9bc7090c819f05c1"
            ]
        ]
    },
    {
        "id": "d98ee674bebb7c69",
        "type": "mqtt out",
        "z": "3af82246.3634ae",
        "name": "",
        "topic": "test",
        "qos": "2",
        "retain": "false",
        "respTopic": "",
        "contentType": "",
        "userProps": "",
        "correl": "",
        "expiry": "",
        "broker": "4c2c9454f646f701",
        "x": 570,
        "y": 380,
        "wires": []
    },
    {
        "id": "a10037027640654f",
        "type": "inject",
        "z": "3af82246.3634ae",
        "name": "",
        "props": [
            {
                "p": "payload"
            },
            {
                "p": "topic",
                "vt": "str"
            }
        ],
        "repeat": "",
        "crontab": "",
        "once": false,
        "onceDelay": 0.1,
        "topic": "",
        "payload": "",
        "payloadType": "date",
        "x": 340,
        "y": 380,
        "wires": [
            [
                "d98ee674bebb7c69"
            ]
        ]
    },
    {
        "id": "4c2c9454f646f701",
        "type": "mqtt-broker",
        "name": "",
        "broker": "${MQTT_SERVER}",
        "port": "${MQTT_PORT}",
        "clientid": "${MQTT_CLIENTID_UNIQUE}",
        "autoConnect": true,
        "usetls": false,
        "protocolVersion": "4",
        "keepalive": "60",
        "cleansession": true,
        "birthTopic": "",
        "birthQos": "0",
        "birthPayload": "",
        "birthMsg": {},
        "closeTopic": "",
        "closeQos": "0",
        "closePayload": "",
        "closeMsg": {},
        "willTopic": "",
        "willQos": "0",
        "willPayload": "",
        "willMsg": {},
        "sessionExpiry": ""
    }
]

### Environment

- Node-RED version: v2.2.2
- Node.js version: v16.3.1
- npm version: v8.1.2
- Platform/OS:Windows/Linux
- Browser: Chrome/Edge

**Comment by Steve-Mcl** [comment]
the `dup` flag is really an internal thing the client deals with for handling QoS

Rightly or wrongly, we have not exposed this to node-red as it is more an implementation detail between the client (MQTT.js) and the Broker.

What is your use case? Are you seeing duplicates because of (for example) network instability?

Are you using QoS 2 (you should never get a duplicate with QoS 2)

Info from [mqtt-essentials-part-6-mqtt-quality-of-service-levels](https://www.hivemq.com/blog/mqtt-essentials-part-6-mqtt-quality-of-service-levels/)

> ## Use QoS 1 when …
> You need to get every message and your use case can handle duplicates. QoS level 1 is the most frequently used service level because it guarantees the message arrives at least once but allows for multiple deliveries. Of course, your application must tolerate duplicates and be able to process them accordingly.
You can’t bear the overhead of QoS 2. QoS 1 delivers messages much faster than QoS 2.
> ## Use QoS 2 when …
> It is critical to your application to receive all messages exactly once. This is often the case if a duplicate delivery can harm application users or subscribing clients. Be aware of the overhead and that the QoS 2 interaction takes more time to complete.

In short, `dup` is a typically client (mqtt.js) implementation for managing QoS 
You will typically not see these unless the broker/client is QoS 1 and something happened between them where the message was not acknowledged by the client (i.e. mqtt.js, i.e. not node-red)

Depending on your "use case", you _could_ utelise MQTT V5 userProperties. e.g. you could add something to `userProperies` (like "UUID": "a-unique-value" or "my-dup": true) to the published `msg` (mqtt-out) and the subscriber (mqtt-in) will receive this.

**Comment by ledat1011** [comment]
I just want to get duplicate flag information from `mqtt-in` when a message arrives

**Comment by Steve-Mcl** [comment]
@knolleary do you have any thoughts on this request please?

**Comment by dceejay** [comment]
As you say this has always (AFAIK) been an internal implementation detail of how the transport implements QoS and has never been exposed at the API level.

**Comment by ledat1011** [comment]
@Steve-Mcl  I just found the method to handle subscribe message from broker, and found this line of code :
https://github.com/node-red/node-red/blob/49e69a54bd9857264fff652309c8c32d309833ac/packages/node_modules/%40node-red/nodes/core/network/10-mqtt.js#L208

This is what will be returned to the user.
you just add this property to the object msg like this 
var msg = {topic:topic, payload:payload, qos:packet.qos, retain:packet.retain ,`dup: packet.dup`};
to get dupflag information for user.

**Comment by knolleary** [comment]
In our years of experience with mqtt, the dup flag has rarely ever proved useful.

All it tells you is that you *might* have seen the message before. If that is the case, you still need to decide if it is a message you have or haven't seen before so you can decide what to do.

For some applications, duplicate messages aren't a problem. For some they are - in which case they should be using qos1 which allows duplicates to happen.

**Comment by ledat1011** [comment]
@knolleary thanks for your answer, please let me know in the future, will you include dup flag in the message ?

**Comment by knolleary** [comment]
Why do you need to use it? As I described, in our experience, the dup flag provides very little benefit.

It would of course be simple to add. But I would rather help you understand why exposing it wouldn't be as helpful as you may think.

**Comment by tfranssen** [comment]
DUP flag only works if the messageId is kept the same at the sender side. This is a typical QoS 1 thing to do. 

Any progress on this issue?
